pkg/operation/botanist/resources.go

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package botanist
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/gardener/gardener/extensions/pkg/controller"
```

Comment:
I wasn't aware about this. Is this documented somewhere? I would very much like to be able to avoid such "forbidden" dependencies myself, what's the best way to go about it besides feedback during a review?

Reply:
@tim-ebert When taking a second look, I am not really sure how to do this, and it's actually why I did it like this in the first place. The reconciliation flow calls the newly introduced `botanist.DeployReferencedResources` and `botanist.DestoryManagedResources` functions. They have to be in `botanist` I think, similarly to most other functions called from the flow. If I move them to `extensions/pkg/controller`, then the dependencies to that package will pop-up in `gardenlet/controller/shoot`, which is not much better. These 2 functions in turn call `controller.CreateManagedResourceFromUnstructured` and `controller.DeleteManagedResources` to deploy referenced resources as a managed resource, as requested by @rfranzke. So I would need to move the entire `managedresources.go` from `extensions/pkg/controller` to somewhere under `pkg`, and I am not sure this is such a great idea. Please advise.

Reply:
> I wasn't aware about this. Is this documented somewhere? I would very much like to be able to avoid such "forbidden" dependencies myself, what's the best way to go about it besides feedback during a review?

No, it's not documented anywhere, as the gardener-extensions library got only merged into g/g recently, and we probably have to learn first how to deal with this new structure. See also https://github.com/gardener/gardener/pull/2251#discussion_r435203263 where I just yesterday made the same comment. If you like you can certainly start a new document under the `docs` folder that describes such things, that would certainly be helpful, also for outside contributors. Though, generally, I'm afraid we'll never reach the situation where everything is perfectly documented because the code structure isn't static, so there such suggestions might occur in the future as well, I assume.

> So I would need to move the entire managedresources.go from extensions/pkg/controller to somewhere under pkg, and I am not sure this is such a great idea. Please advise.

Hm, I don't know of anything that would speak against moving it from `extensions/pkg` to `pkg`. Now that the Gardener core itself is also more and more relying on the `ManagedResource`s it probably is fair to move it.

Reply:
There are probably quite a number of external dependencies (in extensions, etc.) to these `controller` functions for dealing with managed resources. If I move them to `pkg`, all of that would need to be adapted next time Gardener is revendored. Is this ok?
Also, any particular place under `pkg` that you would suggest? None of the existing packages seems like a great fit to me. Perhaps a new package under `utils`, e.g. `utils/managedresources`?

Reply:
Another thing: in `managedresources.go` there are a few references to other functions from `extensions/pkg`, that are used also elsewhere. I will have to move them as well. I am afraid that may trigger a "chain reaction" of moving. Hopefully this is ok.

Reply:
Yes, something like `pkg/utils/managedresources` sounds good to me. As for the adaptations of the package paths in extensions, it should be okay as extensions have to actively revendor to be affected. We can probably make them aware by adding a `action developer` release note to this PR.
Regarding the other functions, I'm not exactly sure which of them you mean, but probably it's okay as well to move them and adapt the calls accordingly.
Generally, I think we still have a lot of stuff in `extensions/pkg` that could be moved to `pkg` now, so any small iteration on refactorings like this are appreciated! We simply "copied" the gardener/gardener-extensions repo to gardener/gardener and didn't spend any time on such refactorings yet, unfortunately.

Reply:
OK, I tried to do as suggested. I put that in a separate commit so you could evaluate the impact more easily. I moved most managed resources utility functions to `pkg/utils/managedresources` as suggested, in fact all of them except `RenderChartAndCreateManagedResource` because that one uses `chart.Render` and moving that would result in a much bigger refactoring, which I believe is better left outside the scope of this PR. There are no more dependencies from `pkg` to `pkg/extensions`.